File: app/src/main/java/com/example/realm_template/ApplicationModule.java
package com.example.realm_template;

import android.content.Context;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.realm.Realm;
import io.realm.RealmConfiguration;

@Module
public class ApplicationModule {

    @Provides
    Context provideApplicationContext() {
        return MyApplication.getInstance();
    }

    @Provides
    @Singleton
    RealmConfiguration provideRealmConfiguration() {
        final RealmConfiguration.Builder builder = new RealmConfiguration.Builder()
                .schemaVersion(Migration.SCHEMA_VERSION)
                .migration(new Migration());
        if (BuildConfig.DEBUG) {
            builder.deleteRealmIfMigrationNeeded();
        }
        return builder.build();
    }

    @Provides
    Realm provideDefaultRealm(RealmConfiguration config) {
        return Realm.getInstance(config);
    }
}

File: kotlin_app/src/main/kotlin/com/example/realm_kotlin_template/ApplicationModule.kt
package com.example.realm_kotlin_template

import android.content.Context

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import io.realm.Realm
import io.realm.RealmConfiguration

@Module
class ApplicationModule {

    @Provides
    @Singleton
    internal fun provideRealmConfiguration(): RealmConfiguration {
        val builder = RealmConfiguration.Builder()
                .schemaVersion(Migration.SCHEMA_VERSION)
                .migration(Migration())
        if (BuildConfig.DEBUG) {
            builder.deleteRealmIfMigrationNeeded()
        }
        return builder.build()
    }

    @Provides
    internal fun provideDefaultRealm(config: RealmConfiguration): Realm {
        return Realm.getInstance(config)
    }
}
